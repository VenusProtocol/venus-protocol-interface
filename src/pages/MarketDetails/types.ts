export interface Stat {
  label: string;
  value: string | number;
}
