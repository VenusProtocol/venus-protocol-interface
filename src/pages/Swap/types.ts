import { SwapDirection, Token } from 'types';

export interface FormValues {
  fromToken: Token;
  fromTokenAmountTokens: string;
  toToken: Token;
  toTokenAmountTokens: string;
  direction: SwapDirection;
}

export type FormError =
  | 'INVALID_FROM_TOKEN_AMOUNT'
  | 'FROM_TOKEN_AMOUNT_HIGHER_THAN_USER_BALANCE'
  | 'FROM_TOKEN_AMOUNT_HIGHER_THAN_WALLET_SPENDING_LIMIT'
  | 'WRAPPING_UNSUPPORTED'
  | 'UNWRAPPING_UNSUPPORTED'
  | 'PRICE_IMPACT_TOO_HIGH';
