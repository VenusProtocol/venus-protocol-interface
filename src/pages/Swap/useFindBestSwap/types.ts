import {
  Currency as PSCurrency,
  Trade as PSTrade,
  TradeType as PSTradeType,
} from '@pancakeswap/sdk/dist/index.js';
import { Token } from 'types';

import { Swap, SwapDirection } from '../types';

export interface UseFindBestSwapInput {
  fromToken: Token;
  toToken: Token;
  direction: SwapDirection;
  fromTokenAmountTokens?: string;
  toTokenAmountTokens?: string;
}

export type UseFindBestSwapOutput = Swap | undefined;

export interface FormatToSwapInput {
  input: UseFindBestSwapInput;
  trade: PSTrade<PSCurrency, PSCurrency, PSTradeType>;
}

export type FormatToSwapOutput = Swap;
