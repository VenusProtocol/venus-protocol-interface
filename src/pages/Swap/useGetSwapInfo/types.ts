import {
  Currency as PSCurrency,
  Trade as PSTrade,
  TradeType as PSTradeType,
} from '@pancakeswap/sdk/dist/index.js';
import { Swap, SwapDirection, Token } from 'types';

export interface UseGetSwapInfoInput {
  fromToken: Token;
  toToken: Token;
  direction: SwapDirection;
  fromTokenAmountTokens?: string;
  toTokenAmountTokens?: string;
}

export type SwapError = 'INSUFFICIENT_LIQUIDITY' | 'WRAPPING_UNWRAPPING_UNSUPPORTED';

export interface FormatToSwapInput {
  input: UseGetSwapInfoInput;
  trade: PSTrade<PSCurrency, PSCurrency, PSTradeType>;
}

export type FormatToSwapOutput = Swap;
