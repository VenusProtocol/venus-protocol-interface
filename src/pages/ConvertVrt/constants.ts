import { getToken } from 'utilities';

export const VRT_ID = 'vrt';
export const VRT_DECIMAL = getToken(VRT_ID).decimals;
