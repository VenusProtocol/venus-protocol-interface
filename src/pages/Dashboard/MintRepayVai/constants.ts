export const VAI_SYMBOL = 'vai';
