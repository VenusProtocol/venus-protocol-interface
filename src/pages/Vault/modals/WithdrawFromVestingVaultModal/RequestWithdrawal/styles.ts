import { css } from '@emotion/react';
import { useTheme } from '@mui/material';

export const useStyles = () => {
  const theme = useTheme();

  return {
    displayWithdrawalRequestListButton: css`
      margin: ${theme.spacing(3, 'auto', -3)};
    `,
  };
};
