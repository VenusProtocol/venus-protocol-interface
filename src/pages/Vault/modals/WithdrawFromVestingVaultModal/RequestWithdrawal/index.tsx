/** @jsxImportSource @emotion/react */
import React, { useContext, useMemo } from 'react';
import BigNumber from 'bignumber.js';

import { AuthContext } from 'context/AuthContext';
import { TOKENS } from 'constants/tokens';
import { TokenId } from 'types';
import { getToken } from 'utilities';
import { useTranslation } from 'translation';
import {
  useGetXvsVaultPoolInfo,
  useGetXvsVaultUserInfo,
  useRequestWithdrawalFromXvsVault,
  useGetXvsVaultLockedDeposits,
} from 'clients/api';
import { ConnectWallet, Spinner, TextButton } from 'components';
import TransactionForm, { ITransactionFormProps } from '../../../TransactionForm';
import { useStyles } from './styles';

export interface RequestWithdrawalProps {
  stakedTokenId: TokenId;
  poolIndex: number;
  handleClose: () => void;
  handleDisplayWithdrawalRequestList: () => void;
}

const RequestWithdrawal: React.FC<RequestWithdrawalProps> = ({
  stakedTokenId,
  poolIndex,
  handleDisplayWithdrawalRequestList,
  handleClose,
}) => {
  const { account } = useContext(AuthContext);
  const stakedToken = getToken(stakedTokenId);
  const { t } = useTranslation();
  const styles = useStyles();

  const {
    mutateAsync: requestWithdrawalFromXvsVault,
    isLoading: isRequestingWithdrawalFromXvsVault,
  } = useRequestWithdrawalFromXvsVault();

  const {
    data: xvsVaultUserLockedDeposits = [],
    isLoading: isGetXvsVaultUserLockedDepositsLoading,
  } = useGetXvsVaultLockedDeposits(
    {
      poolIndex,
      rewardTokenAddress: TOKENS.xvs.address,
      accountAddress: account?.address || '',
    },
    {
      placeholderData: [],
      enabled: !!account?.address,
    },
  );

  const { data: xvsVaultUserInfo, isLoading: isGetXvsVaultUserInfoLoading } =
    useGetXvsVaultUserInfo(
      {
        poolIndex,
        rewardTokenAddress: TOKENS.xvs.address,
        accountAddress: account?.address || '',
      },
      {
        enabled: !!account?.address,
      },
    );

  const requestableWei = useMemo(() => {
    if (!xvsVaultUserInfo?.stakedAmountWei) {
      return new BigNumber(0);
    }

    // Subtract sum of all active withdrawal requests amounts to amount of
    // tokens staked by user
    const pendingLockedDepositsSum = xvsVaultUserLockedDeposits.reduce(
      (acc, xvsVaultUserLockedDeposit) => acc.plus(xvsVaultUserLockedDeposit.amountWei),
      new BigNumber(0),
    );
    return xvsVaultUserInfo.stakedAmountWei.minus(pendingLockedDepositsSum);
  }, [JSON.stringify(xvsVaultUserLockedDeposits), JSON.stringify(xvsVaultUserInfo)]);

  const { data: xvsVaultPoolInfo, isLoading: isGetXvsVaultPoolInfoLoading } =
    useGetXvsVaultPoolInfo(
      {
        poolIndex,
        rewardTokenAddress: TOKENS.xvs.address,
      },
      {
        enabled: !!account?.address,
      },
    );

  const isInitialLoading =
    isGetXvsVaultPoolInfoLoading ||
    isGetXvsVaultUserInfoLoading ||
    isGetXvsVaultUserLockedDepositsLoading;

  const handleSubmit: ITransactionFormProps['onSubmit'] = async amountWei => {
    const res = await requestWithdrawalFromXvsVault({
      poolIndex,
      // account has to be defined at this stage since we don't display the form
      // if no account is connected
      fromAccountAddress: account?.address || '',
      rewardTokenAddress: TOKENS.xvs.address,
      amountWei,
    });

    // Close modal
    handleClose();

    return res;
  };

  return (
    <ConnectWallet
      message={t(
        'withdrawFromVestingVaultModalModal.requestWithdrawalTab.enableToken.connectWalletMessage',
      )}
    >
      {isInitialLoading || !xvsVaultPoolInfo ? (
        <Spinner />
      ) : (
        <>
          <TransactionForm
            tokenId={stakedTokenId}
            availableTokensLabel={t(
              'withdrawFromVestingVaultModalModal.requestWithdrawalTab.availableTokensLabel',
              { tokenSymbol: stakedToken.symbol },
            )}
            availableTokensWei={requestableWei}
            submitButtonLabel={t(
              'withdrawFromVestingVaultModalModal.requestWithdrawalTab.submitButtonLabel',
            )}
            submitButtonDisabledLabel={t(
              'withdrawFromVestingVaultModalModal.requestWithdrawalTab.submitButtonDisabledLabel',
            )}
            successfulTransactionTitle={t(
              'withdrawFromVestingVaultModalModal.requestWithdrawalTab.successfulTransactionTitle',
            )}
            successfulTransactionDescription={t(
              'withdrawFromVestingVaultModalModal.requestWithdrawalTab.successfulTransactionDescription',
            )}
            lockingPeriodMs={xvsVaultPoolInfo.lockingPeriodMs}
            onSubmit={handleSubmit}
            isSubmitting={isRequestingWithdrawalFromXvsVault}
          />

          <TextButton
            onClick={handleDisplayWithdrawalRequestList}
            css={styles.displayWithdrawalRequestListButton}
          >
            {t(
              'withdrawFromVestingVaultModalModal.requestWithdrawalTab.displayWithdrawalRequestListButton',
            )}
          </TextButton>
        </>
      )}
    </ConnectWallet>
  );
};

export default RequestWithdrawal;
