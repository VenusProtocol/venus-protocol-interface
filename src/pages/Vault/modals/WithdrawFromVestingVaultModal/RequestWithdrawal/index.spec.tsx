import React from 'react';
import { act, waitFor, fireEvent } from '@testing-library/react';
import BigNumber from 'bignumber.js';
import noop from 'noop-ts';

import TEST_IDS from 'constants/testIds';
import MAX_UINT256 from 'constants/maxUint256';
import { TokenId } from 'types';
import en from 'translation/translations/en.json';
import { TOKENS } from 'constants/tokens';
import fakeAddress from '__mocks__/models/address';
import xvsVaultResponses from '__mocks__/contracts/xvsVault';
import {
  getAllowance,
  getXvsVaultLockedDeposits,
  getXvsVaultUserInfo,
  getXvsVaultPoolInfo,
  requestWithdrawalFromXvsVault,
} from 'clients/api';
import formatToLockedDeposit from 'clients/api/queries/getXvsVaultLockedDeposits/formatToLockedDeposit';
import formatToUserInfo from 'clients/api/queries/getXvsVaultUserInfo/formatToUserInfo';
import formatToPoolInfo from 'clients/api/queries/getXvsVaultPoolInfo/formatToPoolInfo';
import renderComponent from 'testUtils/renderComponent';
import RequestWithdrawal from '.';

jest.mock('clients/api');

const fakeStakedTokenId = TOKENS.vai.id as TokenId;
const fakePoolIndex = 6;

describe('pages/Vault/modals/WithdrawFromVestingVaultModal/RequestWithdrawal', () => {
  beforeEach(() => {
    // Mark token as enabled
    (getAllowance as jest.Mock).mockImplementation(() => MAX_UINT256);
    (getXvsVaultLockedDeposits as jest.Mock).mockImplementation(() =>
      xvsVaultResponses.getWithdrawalRequests.map(formatToLockedDeposit),
    );
    (getXvsVaultUserInfo as jest.Mock).mockImplementation(() =>
      formatToUserInfo(xvsVaultResponses.userInfo),
    );
    (getXvsVaultPoolInfo as jest.Mock).mockImplementation(() =>
      formatToPoolInfo(xvsVaultResponses.poolInfo),
    );
  });

  it('renders without crashing', async () => {
    const { getByTestId } = renderComponent(
      <RequestWithdrawal
        stakedTokenId={fakeStakedTokenId}
        poolIndex={fakePoolIndex}
        handleClose={noop}
        handleDisplayWithdrawalRequestList={noop}
      />,
      {
        authContextValue: { account: { address: fakeAddress } },
      },
    );

    await waitFor(() => getByTestId(TEST_IDS.vault.transactionForm.availableTokens));
  });

  it('fetches staked tokens and locking period and displays them correctly', async () => {
    const { getByTestId } = renderComponent(
      <RequestWithdrawal
        stakedTokenId={fakeStakedTokenId}
        poolIndex={fakePoolIndex}
        handleClose={noop}
        handleDisplayWithdrawalRequestList={noop}
      />,
      {
        authContextValue: { account: { address: fakeAddress } },
      },
    );

    await waitFor(() => getByTestId(TEST_IDS.vault.transactionForm.availableTokens));

    expect(
      getByTestId(TEST_IDS.vault.transactionForm.availableTokens).textContent,
    ).toMatchSnapshot();
    expect(getByTestId(TEST_IDS.vault.transactionForm.lockingPeriod).textContent).toMatchSnapshot();
  });

  it('calls handleDisplayWithdrawalRequestList callback when clicking on "Withdrawal request list" button', async () => {
    const handleDisplayWithdrawalRequestListMock = jest.fn();

    const { getByText } = renderComponent(
      <RequestWithdrawal
        stakedTokenId={fakeStakedTokenId}
        poolIndex={fakePoolIndex}
        handleClose={noop}
        handleDisplayWithdrawalRequestList={handleDisplayWithdrawalRequestListMock}
      />,
      {
        authContextValue: { account: { address: fakeAddress } },
      },
    );

    await waitFor(() =>
      getByText(
        en.withdrawFromVestingVaultModalModal.requestWithdrawalTab
          .displayWithdrawalRequestListButton,
      ),
    );

    fireEvent.click(
      getByText(
        en.withdrawFromVestingVaultModalModal.requestWithdrawalTab
          .displayWithdrawalRequestListButton,
      ),
    );

    await waitFor(() => expect(handleDisplayWithdrawalRequestListMock).toHaveBeenCalledTimes(1));
  });

  it('lets user request a withdrawal and calls handleClose callback on success', async () => {
    const handleCloseMock = jest.fn();

    const { getByTestId, getByText } = renderComponent(
      <RequestWithdrawal
        stakedTokenId={fakeStakedTokenId}
        poolIndex={fakePoolIndex}
        handleClose={handleCloseMock}
        handleDisplayWithdrawalRequestList={noop}
      />,
      {
        authContextValue: { account: { address: fakeAddress } },
      },
    );

    await waitFor(() => getByTestId(TEST_IDS.vault.transactionForm.availableTokens));

    const fakeValueTokens = '10';

    // Enter amount in input
    act(() => {
      fireEvent.change(getByTestId(TEST_IDS.vault.transactionForm.tokenTextField), {
        target: { value: fakeValueTokens },
      });
    });

    // Submit form
    await waitFor(() =>
      expect(
        getByText(en.withdrawFromVestingVaultModalModal.requestWithdrawalTab.submitButtonLabel),
      ),
    );

    const submitButton = getByText(
      en.withdrawFromVestingVaultModalModal.requestWithdrawalTab.submitButtonLabel,
    ).closest('button') as HTMLButtonElement;

    fireEvent.click(submitButton);

    const fakeWeiSubmitted = new BigNumber(fakeValueTokens).multipliedBy(new BigNumber(10).pow(18));

    await waitFor(() => expect(requestWithdrawalFromXvsVault).toHaveBeenCalledTimes(1));
    expect(requestWithdrawalFromXvsVault).toHaveBeenCalledWith({
      amountWei: fakeWeiSubmitted,
      fromAccountAddress: fakeAddress,
      poolIndex: fakePoolIndex,
      rewardTokenAddress: TOKENS.xvs.address,
    });

    await waitFor(() => expect(handleCloseMock).toHaveBeenCalledTimes(1));
  });
});
