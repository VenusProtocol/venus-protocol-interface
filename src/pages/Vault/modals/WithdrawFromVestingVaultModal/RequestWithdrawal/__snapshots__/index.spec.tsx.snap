// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/Vault/modals/WithdrawFromVestingVaultModal/RequestWithdrawal fetches staked tokens and locking period and displays them correctly 1`] = `"Requestable VAI24 VAI"`;

exports[`pages/Vault/modals/WithdrawFromVestingVaultModal/RequestWithdrawal fetches staked tokens and locking period and displays them correctly 2`] = `"Locking period3 minutes"`;
